tests/pytests/integration/cli/test_salt_call.py

```diff
@@ -429,3 +429,63 @@ def test_local_salt_call_no_function_no_retcode(salt_call_cli):
     assert "test" in ret.data
     assert ret.data["test"] == "'test' is not available."
     assert "test.echo" in ret.data
+
+
+def test_state_highstate_custom_grains(salt_master, salt_minion_factory):
+    """
+    This test ensure that custom grains in salt://_grains are loaded before pillar compilation
+    to ensure that any use of custom grains in pillar files are available, this implies that
+    a sync of grains occurs before loading the regular /etc/salt/grains or configuration file
+    grains, as well as the usual grains.
+
+    Note: cannot use salt_minion and salt_call_cli, since these will be loaded before
+    the pillar and custom_grains files are written, hence using salt_minion_factory.
+    """
+    pillar_top_sls = """
+    base:
+      '*':
+        - defaults
+        """
+
+    pillar_defaults_sls = """
+    mypillar: "{{ grains['custom_grain'] }}"
+    """
+
+    salt_top_sls = """
+    base:
+      '*':
+        - test
+        """
+
+    salt_test_sls = """
+    "donothing":
+      test.nop: []
+    """
+
+    salt_custom_grains_sls = """
+    def main():
+        return {'custom_grain': 'test_value'}
+    """
+    assert salt_master.is_running()
+    with salt_minion_factory.started():
```

Comment:
Since this is an integration test, the minion from `salt_minion_factory` should be running and under `salt_minion`, similar to `salt_master`...

Did you find the need to do this `with` call explicitly?

Reply:
No this was C & P from a test (might be the conftest.py) I was using as an example.

Reply:
> And actually I will leave it the way it is, since I need to create the salt-minion, so that I can make the _grains sync and before the pillar compile. Did it the other way with the built-in salt-minion and ran into chicken and egg problems, this way I get to control just when the egg is created.

In that case, create a separate one time minion just for this?